### DISCUSSION TITLE
consistency of 'repository.url' between event types

### DISCUSSION THREAD

**Discussion by McFoggy** [community/community#24448] [Apps, API and Webhooks]
<p>Depending on which event you listen to, the attribute ‘repository.url’ is different:</p>
<ul>
<li>for <strong>push</strong> events it is equal to the ‘html_url’ attribute &amp; points to the html repository URL not the api URI of the repository</li>
</ul>
<p>    “url”: “<a href="https://github.com/McFoggy/github-api-events" target="_blank" rel="nofollow noopener">https://github.com/McFoggy/github-api-events</a>”</p>
<ul>
<li>for <strong>create</strong>  &amp; other events it is the api URI of the repository and not the html URL.</li>
</ul>
<p>    “url”: “<a href="https://api.github.com/repos/McFoggy/github-api-events" target="_blank" rel="nofollow noopener">https://api.github.com/repos/McFoggy/github-api-events</a>”, </p>
<p>If we look at the repository payload in the API v3 documentation (<a href="https://developer.github.com/v3/repos/#response" rel="nofollow noopener">https://developer.github.com/v3/repos/#response</a>) it should point to the API URL.</p>
<p>Is it for historical reasons or is it a bug?</p>
<p>Is it consistent between <a href="http://github.com" rel="nofollow noopener">github.com</a> &amp; private on premise github installations?</p>
<p>Are there other known inconsistencies?</p>
<p>It is important to me because I am developping software &amp; plugins that target both type of installations.</p>
<p>In both payload below ‘repository.url’ attributes are differents:</p>
<p><em>Push event payload</em></p>
<pre><code>**&lt;u&gt; "repository": {
    "id": 178348886,
    "node_id": "MDEwOlJlcG9zaXRvcnkxNzgzNDg4ODY=",
    "name": "github-api-events",
    "full_name": "McFoggy/github-api-events",
    "private": false,
    "owner": {
      "name": "McFoggy",
      ...
    },
    "html_url": "https://github.com/McFoggy/github-api-events",
    "url": "https://github.com/McFoggy/github-api-events",
    "forks_url": "https://api.github.com/repos/McFoggy/github-api-events/forks",
    ...
}&lt;/u&gt;**
</code></pre>
<p><em>Create &amp; other events payload</em></p>
<pre><code>**&lt;u&gt; "repository": {
    "id": 178348886,
    "node_id": "MDEwOlJlcG9zaXRvcnkxNzgzNDg4ODY=",
    "name": "github-api-events",
    "full_name": "McFoggy/github-api-events",
    "private": false,
    "owner": {
      "login": "McFoggy",
      "id": 1119660,
      ...
    },
    "html_url": "https://github.com/McFoggy/github-api-events",
    "url": "https://api.github.com/repos/McFoggy/github-api-events",
    "forks_url": "https://api.github.com/repos/McFoggy/github-api-events/forks",
    ...
}&lt;/u&gt;**
</code></pre>
<p>I collected complete payloads of webhooks in an example project: <a href="https://github.com/McFoggy/github-api-events/blob/master/README.md" rel="nofollow noopener">https://github.com/McFoggy/github-api-events/blob/master/README.md</a></p>

**Comment by McFoggy** [comment]
<p><em>PING</em></p>
<p>Does someone know how the on premise enterprise edition reacts in this case?</p>
<p>Is the payload the same as the <a href="http://github.com" rel="nofollow noopener">github.com</a> API or does it conform to the documentation?</p>
<p>Thank for any help that could be provided.</p>

**Comment by lee-dohm** [comment]
<p>It appears to be a bug but I haven’t been able to confirm that with the development team at this time. I don’t know if there are any other inconsistencies, no. I can’t promise if or when this would be fixed, unfortunately.</p>
<p>My recommendation, at least for now, would be to use the <a href="https://developer.github.com/v3/repos/#get" rel="nofollow noopener">documented v3 REST API URLs for retrieving information</a> rather than relying on the API URLs in the event objects themselves.</p>
<p>I hope that helps!</p>

**Comment by webknjaz** [comment]
<p>Yeah, that’s what I ended up doing as well. But a proper fix would be great to have :slight_smile:</p>

**Comment by McFoggy** [comment, marked as answer]
<blockquote>
<p><a class="mention" href="/lee-dohm">@lee-dohm</a> wrote:</p>
</blockquote>
<blockquote>
<p>My recommendation, at least for now, would be to use the <a href="https://developer.github.com/v3/repos/#get" rel="nofollow noopener">documented v3 REST API URLs for retrieving information</a> rather than relying on the API URLs in the event objects themselves.</p>
</blockquote>
<p>But when you have a webservice/application/integration that listens to a webhook event  ; in order to call back the server which fired the webhook you have to use the ‘repository.url’ field which contains a wrong value :wink:</p>
<p>I’ll implement some logic (receiver side of the webhook) to detect which field contains the good value.</p>

**Comment by McFoggy** [comment]
<p><a class="mention" href="/lee-dohm">@lee-dohm</a> any news from the dev team?</p>

**Comment by lee-dohm** [comment]
<p>Nope, my recommendation still stands. Use the <a href="https://developer.github.com/v3/" rel="nofollow noopener">documented endpoints</a>. Or you could convert to using the GraphQL v4 API which only has one endpoint to use for everything.</p>
